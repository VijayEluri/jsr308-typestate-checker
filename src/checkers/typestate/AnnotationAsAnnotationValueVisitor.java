package checkers.typestate;

import checkers.util.AnnotationUtils;
import checkers.nullness.quals.Nullable;

import javax.lang.model.element.*;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Types;
import java.util.List;

/**
 * A visitor which, for the value of a parameter of an annotation, returns the corresponding {@link AnnotationMirror}
 * if the value is an annotation, or null otherwise.
 * @author Adam Warski (adam at warski dot org)
*/
public class AnnotationAsAnnotationValueVisitor implements AnnotationValueVisitor</*@Nullable*/ AnnotationMirror, Void> {
    private final AnnotationUtils annotationUtils;
    private final Types types;

    public AnnotationAsAnnotationValueVisitor(AnnotationUtils annotationUtils, Types types) {
        this.annotationUtils = annotationUtils;
        this.types = types;
    }

    public /*@Nullable*/ AnnotationMirror visit(AnnotationValue av, Void aVoid) { return null; }
    public /*@Nullable*/ AnnotationMirror visit(AnnotationValue av) { return null; }
    public /*@Nullable*/ AnnotationMirror visitBoolean(boolean b, Void aVoid) { return null; }
    public /*@Nullable*/ AnnotationMirror visitByte(byte b, Void aVoid) { return null; }
    public /*@Nullable*/ AnnotationMirror visitChar(char c, Void aVoid) { return null; }
    public /*@Nullable*/ AnnotationMirror visitDouble(double d, Void aVoid) { return null; }
    public /*@Nullable*/ AnnotationMirror visitFloat(float f, Void aVoid) { return null; }
    public /*@Nullable*/ AnnotationMirror visitInt(int i, Void aVoid) { return null; }
    public /*@Nullable*/ AnnotationMirror visitLong(long i, Void aVoid) { return null; }
    public /*@Nullable*/ AnnotationMirror visitShort(short s, Void aVoid) { return null; }
    public /*@Nullable*/ AnnotationMirror visitString(String s, Void aVoid) { return null; }
    public /*@Nullable*/ AnnotationMirror visitEnumConstant(VariableElement c, Void aVoid) { return null; }
    public /*@Nullable*/ AnnotationMirror visitAnnotation(AnnotationMirror a, Void aVoid) { return null; }
    public /*@Nullable*/ AnnotationMirror visitArray(List<? extends AnnotationValue> vals, Void aVoid) { return null; }
    public /*@Nullable*/ AnnotationMirror visitUnknown(AnnotationValue av, Void aVoid) { return null; }

    // We're only interested in class values, which are additionally annotations.
    public /*@Nullable*/ AnnotationMirror visitType(TypeMirror t, Void aVoid) {
        if (types.asElement(t).getKind() == ElementKind.ANNOTATION_TYPE) {
            return annotationUtils.fromName(t.toString());
        } else {
            return null;
        }
    }
}
