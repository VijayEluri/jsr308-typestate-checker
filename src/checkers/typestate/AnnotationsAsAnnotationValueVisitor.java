package checkers.typestate;

import checkers.util.AnnotationUtils;

import javax.lang.model.element.*;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Types;
import java.util.List;
import java.util.Set;

/**
 * A visitor which, for the value of a parameter of an annotation,
 * returns the corresponding set of {@link javax.lang.model.element.AnnotationMirror}
 * if the value is a set of annotations, or null otherwise.
 * @author Adam Warski (adam at warski dot org)
*/
public class AnnotationsAsAnnotationValueVisitor implements AnnotationValueVisitor<Set<AnnotationMirror>, Void> {
    private final AnnotationAsAnnotationValueVisitor annotationAsAnnotationValueVisitor;

    public AnnotationsAsAnnotationValueVisitor(AnnotationUtils annotationUtils, Types types) {
        this.annotationAsAnnotationValueVisitor = new AnnotationAsAnnotationValueVisitor(annotationUtils, types);
    }

    public Set<AnnotationMirror> visit(AnnotationValue av, Void aVoid) { return null; }
    public Set<AnnotationMirror> visit(AnnotationValue av) { return null; }
    public Set<AnnotationMirror> visitBoolean(boolean b, Void aVoid) { return null; }
    public Set<AnnotationMirror> visitByte(byte b, Void aVoid) { return null; }
    public Set<AnnotationMirror> visitChar(char c, Void aVoid) { return null; }
    public Set<AnnotationMirror> visitDouble(double d, Void aVoid) { return null; }
    public Set<AnnotationMirror> visitFloat(float f, Void aVoid) { return null; }
    public Set<AnnotationMirror> visitInt(int i, Void aVoid) { return null; }
    public Set<AnnotationMirror> visitLong(long i, Void aVoid) { return null; }
    public Set<AnnotationMirror> visitShort(short s, Void aVoid) { return null; }
    public Set<AnnotationMirror> visitString(String s, Void aVoid) { return null; }
    public Set<AnnotationMirror> visitEnumConstant(VariableElement c, Void aVoid) { return null; }
    public Set<AnnotationMirror> visitAnnotation(AnnotationMirror a, Void aVoid) { return null; }
    public Set<AnnotationMirror> visitUnknown(AnnotationValue av, Void aVoid) { return null; }
    public Set<AnnotationMirror> visitType(TypeMirror t, Void aVoid) { return null; }

    public Set<AnnotationMirror> visitArray(List<? extends AnnotationValue> vals, Void aVoid) {
		Set<AnnotationMirror> result = AnnotationUtils.createAnnotationSet();
		for (AnnotationValue val : vals) {
			AnnotationMirror annotation = annotationAsAnnotationValueVisitor.visit(val);

			if (annotation != null) {
				result.add(annotation);
			}
		}

		return result;
	}
}