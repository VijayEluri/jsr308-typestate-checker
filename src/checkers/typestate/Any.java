package checkers.typestate;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Annotation;

/**
 * A special state annotation, specifying that an object may be in an arbitrary state. As a lack of a state annotation
 * means "any annotation", this is only useful when specifying state transitions, the {@code except} element or the
 * {@code onException} element.
 * @author Adam Warski (adam at warski dot org)
 */
@Retention(RetentionPolicy.RUNTIME)
public @interface Any {
	/**
	 * @return The state of the object after the method in which the annotations is used completes.
	 */
    Class<? extends Annotation> after()			default NoChange.class;

	/**
	 * @return States that this annotation does not cover. If the object will be in a state that is listed here,
	 * the transition specified by this annotation won't take effect, nor will this annotation accept the object.
	 */
	Class<? extends Annotation>[] except()		default {};

	/**
	 * @return The state of the object if the method in which the annotation is used throws an exception.
	 */
    Class<? extends Annotation> onException()	default NoChange.class;
}