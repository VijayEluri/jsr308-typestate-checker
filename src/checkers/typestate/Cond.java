package checkers.typestate;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.lang.annotation.ElementType;
import java.lang.annotation.Annotation;

/**
 * Annotation for methods, for which the receiver's state changes depending on the boolean result of the method.
 * @author Adam Warski (adam at warski dot org)
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface Cond {
	/**
	 * @return The state of the object if the annotated method returns true.
	 */
    Class<? extends Annotation> whenTrue()			default NoChange.class;

	/**
	 * @return The state of the object if the annotated method returns false.
	 */
    Class<? extends Annotation> whenFalse()			default NoChange.class;
}
