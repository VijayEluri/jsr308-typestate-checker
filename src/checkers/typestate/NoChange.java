package checkers.typestate;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * @author Adam Warski (adam at warski dot org)
 */
@Retention(RetentionPolicy.RUNTIME)
@Target({})
public @interface NoChange {}