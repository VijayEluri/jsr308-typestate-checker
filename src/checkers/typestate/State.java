package checkers.typestate;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import static java.lang.annotation.ElementType.*;

/**
 * Marks an annotation as a state annotation. Correct usage and placement of such annotations will be checked
 * by the typestate checker.
 * @author Adam Warski (adam at warski dot org)
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ANNOTATION_TYPE)
public @interface State {}
