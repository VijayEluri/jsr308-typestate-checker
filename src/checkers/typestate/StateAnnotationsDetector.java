package checkers.typestate;

import com.sun.source.util.TreePathScanner;
import com.sun.source.util.Trees;
import com.sun.source.tree.*;

import java.util.Set;
import java.util.List;

import checkers.util.AnnotationUtils;
import checkers.types.AnnotatedTypeMirror;
import checkers.types.AnnotatedTypeFactory;

import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeKind;
import javax.annotation.processing.ProcessingEnvironment;

/**
 * Detects all annotations annotated with {@link State}, used on any program element.
 * @author Adam Warski (adam at warski dot org)
 */
public class StateAnnotationsDetector extends TreePathScanner<Void, Set<AnnotationMirror>> {
    private final AnnotationUtils annotationUtils;
    private final ProcessingEnvironment env;
    private final TypestateUtil typestateUtil;
    private final AnnotatedTypeFactory factory;

    public StateAnnotationsDetector(ProcessingEnvironment env, TypestateUtil typestateUtil,
                                    AnnotatedTypeFactory factory) {
        this.annotationUtils = AnnotationUtils.getInstance(env);
        this.env = env;
        this.typestateUtil = typestateUtil;
        this.factory = factory;
    }

    private void addStateAnnotation(AnnotationMirror stateAnnotation, boolean isPure, Set<AnnotationMirror> to) {
        if (!to.contains(stateAnnotation)) {
            // Adding "pure" annotations, that is, without any parameters set.
            if (isPure) {
                to.add(stateAnnotation);
            } else {
                to.add(annotationUtils.fromName(stateAnnotation.getAnnotationType().toString()));
            }
        }
    }

    private void addStateAnnotations(Set<AnnotationMirror> from, Set<AnnotationMirror> to) {
        for (AnnotationMirror annotation : from) {
			boolean isStateAnnotation = typestateUtil.isStateAnnotation(annotation);

            if (isStateAnnotation) {
                addStateAnnotation(annotation, false, to);
            }

			if (typestateUtil.isAnyStateAnnotation(annotation) || isStateAnnotation) {
                // Checking if the annotation doesn't define a transition to another state. If so, adding that state.
                AnnotationMirror afterAnnotation = typestateUtil.getAfterParameterValue(annotation);
                if (afterAnnotation != null) {
                    addStateAnnotation(afterAnnotation, true, to);
                }

				// Checking if the annotation doesn't define an exception-state. If so, adding that state.
                AnnotationMirror exceptionAnnotation = typestateUtil.getExceptionParameterValue(annotation);
                if (exceptionAnnotation != null) {
                    addStateAnnotation(exceptionAnnotation, true, to);
                }

				// And, in case of an any-state annotation, if the "except" element is set and if it contains
				// state annotations.
				List<AnnotationMirror> exceptAnnotations = typestateUtil.getExceptParameterValue(annotation);
				if (exceptAnnotations != null) {
					for (AnnotationMirror exceptAnnotation : exceptAnnotations) {
					    if (typestateUtil.isStateAnnotation(exceptAnnotation)) {
							addStateAnnotation(exceptAnnotation, true, to);
						}
					}
				}
			}
        }
    }

    public Void visitMethodInvocation(MethodInvocationTree node, Set<AnnotationMirror> stateAnnotations) {
        // Adding all annotations that may be present on the invoked method return type, parameters and receiver.
        AnnotatedTypeMirror.AnnotatedExecutableType invocationType = factory.methodFromUse(node);

        addStateAnnotations(invocationType.getReceiverType().getAnnotations(), stateAnnotations);
        addStateAnnotations(invocationType.getReturnType().getAnnotations(), stateAnnotations);
        for (AnnotatedTypeMirror parameter : invocationType.getParameterTypes()) {
            addStateAnnotations(parameter.getAnnotations(), stateAnnotations);
        }

        return super.visitMethodInvocation(node, stateAnnotations);
    }

    public Void visitMethod(MethodTree node, Set<AnnotationMirror> stateAnnotations) {
        // Adding all annotations that may be present on the declared method return type, parameters and receiver.
        AnnotatedTypeMirror.AnnotatedExecutableType methodType = factory.getAnnotatedType(node);

        addStateAnnotations(methodType.getReceiverType().getAnnotations(), stateAnnotations);
        addStateAnnotations(methodType.getReturnType().getAnnotations(), stateAnnotations);
        for (AnnotatedTypeMirror parameter : methodType.getParameterTypes()) {
            addStateAnnotations(parameter.getAnnotations(), stateAnnotations);
        }

        return super.visitMethod(node, stateAnnotations);   
    }

    public Void visitAnnotation(AnnotationTree node, Set<AnnotationMirror> stateAnnotations) {
        TypeMirror nodeTypeMirror = Trees.instance(env).getTypeMirror(getCurrentPath());

        if (nodeTypeMirror.getKind() == TypeKind.DECLARED) {
            DeclaredType dt = (DeclaredType) nodeTypeMirror;
            if (typestateUtil.isStateAnnotation(dt)) {
                addStateAnnotation(annotationUtils.fromName(dt.toString()), true, stateAnnotations);
            }
        } else {
            throw new UnsupportedOperationException("Unsupported type of annotation: "
                    + nodeTypeMirror.getKind() + ".");
        }

        return super.visitAnnotation(node, stateAnnotations);
    }
}
