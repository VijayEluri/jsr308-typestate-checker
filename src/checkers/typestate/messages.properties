receiver.in.wrong.state=The method receiver '%s' is in the wrong state. Expected: %s, actual: %s.
parameter.in.wrong.state=The parameter '%s' is in the wrong state. Expected: %s, actual: %s.