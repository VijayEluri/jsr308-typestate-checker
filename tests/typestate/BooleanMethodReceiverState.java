import checkers.typestate.State;
import checkers.typestate.NoChange;

/**
 * @author Adam Warski (adam at warski dot org)
 */
public class BooleanMethodReceiverState {
	@State public static @interface State0 { public abstract Class<?> afterTrue() default NoChange.class; public abstract Class<?> afterFalse() default NoChange.class; }
    @State public static @interface State1 { }
    @State public static @interface State2 { }

    public static class Helper {
        public void onlyInState1() /*@State1*/ { }
        public void onlyInState2() /*@State2*/ { }

		public boolean modify() /*@State0(afterTrue=State1.class, afterFalse=State2.class)*/ { return true; }
    }

    public void testOk1(@State0 Helper h) {
        if (h.modify()) {
			h.onlyInState1();
		} else {
			h.onlyInState2();
		}
    }

	public void testOk2(@State0 Helper h) {
        if (h.modify() == true) {
			h.onlyInState1();
		} else {
			h.onlyInState2();
		}
    }

	public void testOk3(@State0 Helper h) {
        if (h.modify() == false) {
			h.onlyInState2();
		} else {
			h.onlyInState1();
		}
    }

	public void testOk4(@State0 Helper h) {
        if (h.modify() != true) {
			h.onlyInState2();
		} else {
			h.onlyInState1();
		}
    }

	public void testOk5(@State0 Helper h) {
        if (!(h.modify() == false)) {
			h.onlyInState1();
		} else {
			h.onlyInState2();
		}
    }

    public void testError(@State0 Helper h) {
        if (h.modify()) {
			h.onlyInState2();
		} else {
			h.onlyInState1();
		}
    }
}