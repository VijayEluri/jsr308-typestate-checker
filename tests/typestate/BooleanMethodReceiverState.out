:63: (receiver.in.wrong.state)
:61: (receiver.in.wrong.state)
